src/v/datalake/catalog_schema_manager.cc

```diff
@@ -230,7 +235,10 @@ catalog_schema_manager::get_table_info(
     }
     const auto& table = load_res.value();
 
-    auto cur_schema = table.get_schema(table.current_schema_id);
+    const auto* cur_schema = desired_type.has_value()
+                               ? table.get_equivalent_schema(
```

Comment:
I wonder if there are perf implications here? Might be good to check what the impact here is and if there some other mechanism we want to have here (or does this get cached per schema ID in the translation path - I will look).

Reply:
Yeah there are other options like using a fingerprint to quickly sort out mismatches. Hopefully it's all behind a cache and this won't matter

Reply:
Just to clarify, the caching I was referring to is pretty indirect -- it's caching at the level of the record multiplexer, such that each translation of an offset range will have a shared map of parquet writers that's more or less keyed by schema ID. So we will need to do this lookup once per record schema per translation.

Reply:
Right right. So assuming that:

- the steady state of most translators is a stream of records conforming to the corresponding table's current schema
- the first thing we look up in get_table_info is the current table schema

Then incurring a single additional equivalence check in the common case seems basically fine?

In principle we should only get in the out-of-sync producer state when the schema changes, but I don't have a good sense of how long we might spend there. 

maybe an on-shard cache keyed by some fingerprint is good enough? Vs the additional work of wiring something into the coordinator

Reply:
@andrwng - completely misread your comment 🤦.

> once per record schema per translation

should be totally fine then as written, yeah?

Reply:
Oops I missed the earlier comment. Yeah I wasn't implying that what we have is insufficient, just thought i'd clarify where there is schema reuse. I think it's fine as is (or at least, let's wait and see as the translation scheduling policy crystalizes, though I'd be surprised if it becomes a bottleneck)

I will say, maybe this is worth trying out with some massive schemas

Reply:
no worries. yeah for sure - what i read on mobile was "check per **batch**", so I went off an wrote a pointless benchmark... entirely my bad.

> maybe this is worth trying out with some massive schemas

yeah i can whip something up at some point.

> wait and see...surprised if it becomes a bottleneck

agree